doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+
+First we need to define a few terms:
+
+- _template name_: This is the main identifier of a template. It is used in the command line to choose the template.
+- _template base name_: Unlike the template name, this identifier is purely internal. It is the base name of all mustache files belonging
+  to the template. In most cases this will be the same as the template name.
+- _input keyset_: This keyset contains the data with which the template will be instantiated by the code-generator.
+- _parent key_: The parent key is given in the command line and defines the input keyset (the keys below the parent key).
+- _output name_: The base name for the output files as given in the command line. Suffixes may be appended, if there are multiple output
+  files.
+- _parts_: Each template may consist of multiple parts. Each part corresponds to exactly one mustache file and therefore exactly one output
+  file.
+- _part suffix_: Parts are identified by there part suffix. This suffix is appended to the template base name to obtain the name of the
+  mustache file corresponding to the part. Similarly it is appended to the output name to obtain the output file for this part. The empty
+  string `""` is a valid part suffix, although it is only recommend for templates with a single output file.
+
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing. The file will simple list all keys in our input
+keyset. An example output would be (the same as `kdb ls`):
+
+```
+user/sw/myapp/#0/current
+user/sw/myapp/#0/current/dir0
+user/sw/myapp/#0/current/dir0/subdir0/key0
+user/sw/myapp/#0/current/dir0/subdir0/key1
+user/sw/myapp/#0/current/dir1
+```
+
+### Creating the mustache template
+
+The first thing we need to create is a mustache template. You will need (at least) one template file for each output file. If you make use
+of partials (see [below](#using-partials)), you of course have to write multiple files.
+
+All template files must be located in the [`src/tools/kdb/gen/templates`](/src/tools/kdb/gen/templates) folder and must be named according
+to the scheme `<template base name>.<part suffix>.mustache`.
+
+For our simple example we create the file `src/tools/kdb/gen/templates/example.txt.mustache` with the following content:
+
+```
+{{# keys }}
+{{{ name }}}
+{{/ keys }}
+```
+
+Note: we will not go into detail on how mustache templates work, for more information see e.g.
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+### Creating the supporting class
+
+Since we need to some way of supplying data to our template, we have to create a subclass of `GenTemplate`.
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
```

Comment:
The directory can not be somewhere else?

Reply:
No, because the CMake scripts expects the files in this place.

Reply:
Are the templates then also built-in the binaries?

Reply:
Which binaries?

Right now CMake reads all the `.mustache` and puts their contents into string constants (`static const char *`), because that was the easiest way to use the mustache library. We could also install the template files and read them at runtime, but then we would have to do some trickery to allow running the code-generator without installing first.

Reply:
> Which binaries?

`kdb` (and `kdb-full`, ...)

> Right now CMake reads all the .mustache and puts their contents into string constants 

Would be good to mention this here. Then it is less surprising why we modify files in `src/tools/kdb/gen`

Reply:
I just noticed that I confuse the lines here. For the `.cpp` and `.hpp` files, it doesn't really matter where they are, but putting the files in a different folder might required modifying the CMake files.

I will add some clarifications.